### PR TITLE
[WIP] Collect improvements

### PR DESCRIPTION
**Pull request by dfalbel**

### CODE REVIEW

File: vignettes/articles/read.Rmd [hunk 1]
```diff
@@ -0,0 +1,106 @@
+---
```
Comment (dfalbel) [inline vignettes/articles/read.Rmd:1]: @t-kalinowski I'm leaning towards this interface. Do you have comments?

**Comment by t-kalinowski** [comment]
Attempting to build on Ubuntu I get an error.
```
[ 99%] Building CXX object CMakeFiles/libprotoc.dir/src/google/protobuf/compiler/ruby/ruby_generator.cc.o
[ 99%] Building CXX object CMakeFiles/libprotoc.dir/src/google/protobuf/compiler/subprocess.cc.o
[100%] Building CXX object CMakeFiles/libprotoc.dir/src/google/protobuf/compiler/zip_writer.cc.o
[100%] Linking CXX static library libprotoc.a
[100%] Built target libprotoc
Scanning dependencies of target protoc
[100%] Building CXX object CMakeFiles/protoc.dir/src/google/protobuf/compiler/main.cc.o
[100%] Linking CXX executable protoc
[100%] Built target protoc
** libs
g++ -std=gnu++11 -I"/home/tomasz/opt/R-4.2.1/lib/R/include" -DNDEBUG -I../inst/tp/pb/src -Igenerated -I'/home/tomasz/opt/R-4.2.1/lib/R/site-library/Rcpp/include' -I/usr/local/include   -fpic  -O2 -march=native -mtune=native -flto -c RcppExports.cpp -o RcppExports.o
In file included from ../inst/tp/pb/src/google/protobuf/io/coded_stream.h:130,
                 from generated/event.pb.h:24,
                 from event_writer.h:2,
                 from tfevents_types.h:1,
                 from RcppExports.cpp:4:
../inst/tp/pb/src/google/protobuf/stubs/common.h:44:10: fatal error: absl/strings/string_view.h: No such file or directory
   44 | #include "absl/strings/string_view.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [/home/tomasz/opt/R-4.2.1/lib/R/etc/Makeconf:177: RcppExports.o] Error 1
ERROR: compilation failed for package 'tfevents'
* removing '/home/tomasz/opt/R-4.2.1/lib/R/site-library/tfevents'
Warning in i.p(...) :
  installation of package '/tmp/Rtmp6qd141/file221686f767/tfevents_0.0.0.9000.tar.gz' had non-zero exit status
```
Any ideas for how to fix?

**Comment by dfalbel** [comment]
Maybe submodules haven't been recursively cloned. I'd try to to delete and clone with:

```
git clone --recurse-submodules
```

**Comment by t-kalinowski** [comment]
Still errors:
```
$ git clone --recurse-submodules -j8 https://github.com/mlverse/tfevents
Cloning into 'tfevents'...
remote: Enumerating objects: 1474, done.
remote: Counting objects: 100% (274/274), done.
remote: Compressing objects: 100% (181/181), done.
remote: Total 1474 (delta 164), reused 173 (delta 86), pack-reused 1200
Receiving objects: 100% (1474/1474), 4.11 MiB | 10.90 MiB/s, done.
Resolving deltas: 100% (889/889), done.
Submodule 'src/protobuf' (https://github.com/protocolbuffers/protobuf) registered for path 'inst/tp/pb'
Cloning into '/home/tomasz/r-pkgs/tfevents/inst/tp/pb'...
remote: Enumerating objects: 116202, done.        
remote: Counting objects: 100% (28/28), done.        
remote: Compressing objects: 100% (21/21), done.        
remote: Total 116202 (delta 2), reused 13 (delta 0), pack-reused 116174        
Receiving objects: 100% (116202/116202), 108.05 MiB | 23.19 MiB/s, done.
Resolving deltas: 100% (82402/82402), done.
Submodule path 'inst/tp/pb': checked out '90b73ac3f0b10320315c2ca0d03a5a9b095d2f66'
Submodule 'third_party/benchmark' (https://github.com/google/benchmark.git) registered for path 'inst/tp/pb/third_party/benchmark'
Submodule 'third_party/googletest' (https://github.com/google/googletest.git) registered for path 'inst/tp/pb/third_party/googletest'
Cloning into '/home/tomasz/r-pkgs/tfevents/inst/tp/pb/third_party/benchmark'...
remote: Enumerating objects: 7548, done.        
remote: Counting objects: 100% (84/84), done.        
remote: Compressing objects: 100% (63/63), done.        
remote: Total 7548 (delta 32), reused 58 (delta 17), pack-reused 7464        
Receiving objects: 100% (7548/7548), 2.37 MiB | 5.03 MiB/s, done.
Resolving deltas: 100% (5009/5009), done.
Cloning into '/home/tomasz/r-pkgs/tfevents/inst/tp/pb/third_party/googletest'...
remote: Enumerating objects: 25258, done.        
remote: Counting objects: 100% (167/167), done.        
remote: Compressing objects: 100% (90/90), done.        
remote: Total 25258 (delta 87), reused 117 (delta 67), pack-reused 25091        
Receiving objects: 100% (25258/25258), 11.08 MiB | 11.44 MiB/s, done.
Resolving deltas: 100% (18658/18658), done.
Submodule path 'inst/tp/pb/third_party/benchmark': checked out '5b7683f49e1e9223cf9927b24f6fd3d6bd82e3f8'
Submodule path 'inst/tp/pb/third_party/googletest': checked out '5ec7f0c4a113e2f18ac2c6cc7df51ad6afc24081'
tomasz@horse:~/r-pkgs
$ cd tfevents
tomasz@horse:~/r-pkgs/tfevents (main) 
$ Rscript -e 'remotes::install_local()'
Warning in stats::reshape(values, idvar = "submodule", timevar = "name",  :
  partial argument match of 'v.name' to 'v.names'
'/usr/bin/git' clone --depth 1 --no-hardlinks --recurse-submodules https://github.com/protocolbuffers/protobuf /tmp/RtmpCoe0PL/file37c54efe8896/tfevents/inst/tp/pb
fatal: destination path '/tmp/RtmpCoe0PL/file37c54efe8896/tfevents/inst/tp/pb' already exists and is not an empty directory.
Warning in system(full, intern = TRUE, ignore.stderr = quiet) :
  running command ''/usr/bin/git' clone --depth 1 --no-hardlinks --recurse-submodules https://github.com/protocolbuffers/protobuf /tmp/RtmpCoe0PL/file37c54efe8896/tfevents/inst/tp/pb' had status 128
Error: Failed to install 'tfevents' from local:
  Command failed (128)
Execution halted
```

**Comment by dfalbel** [comment]
It seems that `remotes::install_local()` tries to re-clone stuff. I think `devtools::build()` should work, or `R CMD INSTALL`?

**Comment by dfalbel** [comment]
So, it looks like that `remotes::install_github()` will clone the main branch of the submodules while we are tagging the latest version - which is 21.12. That version doesn't include abseil-cpp as a dependency.

This is likely a bug in `install_github` but it seems a hard one to fix as the commit SHA of submodules is stored in a git object inside the .git directory. However this directory is not exposed when remotes downloads the source code.

I think the easy fix for us is to simply vendor all the protobuf files instead of using submodules.

**Comment by t-kalinowski** [comment]
Hi @dfalbel,

Thanks for putting this together! Looks great :) 

Some minor notes:

-  I'm seeing this error:
```r
> knitr::opts_chunk$set(
+   collapse = TRUE,
+   comment = "#>"
+ )
> library(tfevents)
> temp <- tempfile()
> set_default_logdir(temp)
> log_event(text = "hello world")
> for (i in 1:10) {
+   log_event(hello = log(i + runif(1)))
+ }
> events <- collect_events(temp)
> events
Error: Corrupt x: no names
```

```r
summaries <- collect_events(temp, type = "summary")
summaries
```

- Is this expected? Are summary objects meant to return themselves when subsetted? 
```r
summaries$summary[[1]][[1]][[1]][[1]][[1]]
```


-  Would it be possible to make `as.list()` return the values as R objects directly?
```
as.list(summaries$summary) # instead of sapply(,value)`
``` 

-  Is there anything else a user would do with a `<tfevents_summary_values>` other than extract the value? Maybe we don't need expose the protobuf structure and `value()` function and can just return R values to the user. E.g., `collect_events()` returns a data frame with a list column, each entry an atomic R vector, maybe with different `class()`'s for the different plugin types like `structure(c(...), class = "tfevents_image", dim = c(128,3,3)`. What do you think?

--- 
```r
con <- events_logdir(temp)
collect_events(con, n = 1, type = "scalar")
```
This is great!

**Pull request merged**